kedro-datasets/kedro_datasets/databricks/managed_table_dataset.py

```diff
@@ -0,0 +1,433 @@
+"""``ManagedTableDataSet`` implementation to access managed delta tables
+in Databricks.
+"""
+import logging
+import re
+from dataclasses import dataclass
+from typing import Any, Dict, List, Optional, Union
+
+import pandas as pd
+from kedro.io.core import (
+    AbstractVersionedDataSet,
+    DataSetError,
+    Version,
+    VersionNotFoundError,
+)
+from pyspark.sql import DataFrame, SparkSession
+from pyspark.sql.types import StructType
+from pyspark.sql.utils import AnalysisException, ParseException
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass(frozen=True)
+class ManagedTable:  # pylint: disable=too-many-instance-attributes
+    """Stores the definition of a managed table"""
+
+    # regex for tables, catalogs and schemas
+    _NAMING_REGEX = r"\b[0-9a-zA-Z_-]{1,32}\b"
+    _VALID_WRITE_MODES = ["overwrite", "upsert", "append"]
+    _VALID_DATAFRAME_TYPES = ["spark", "pandas"]
+    database: str
+    catalog: str
+    table: str
+    write_mode: str
+    dataframe_type: str
+    primary_key: Optional[str]
+    owner_group: str
+    partition_columns: Union[str, List[str]]
+    json_schema: StructType
+
+    def __post_init__(self):
+        """Run validation methods if declared.
+        The validation method can be a simple check
+        that raises DataSetError.
+        The validation is performed by calling a function named:
+            `validate_<field_name>(self, value) -> raises DataSetError`
+        """
+        for name in self.__dataclass_fields__.keys():  # pylint: disable=no-member
+            method = getattr(self, f"validate_{name}", None)
+            if method:
+                method()
+
+    def validate_table(self):
+        """validates table name
+
+        Raises:
+            DataSetError:
+        """
+        if not re.fullmatch(self._NAMING_REGEX, self.table):
+            raise DataSetError("table does not conform to naming")
+
+    def validate_database(self):
+        """validates database name
+
+        Raises:
+            DataSetError: If the table name does not conform to naming constraints.
+        """
+        if self.database:
+            if not re.fullmatch(self._NAMING_REGEX, self.database):
+                raise DataSetError("database does not conform to naming")
+
+    def validate_catalog(self):
```

Comment:
These look to be doing the same thing, could we not just take in an argument to determine if it's a table, database or catalog and have 1 method?

Reply:
This is a great push. Managed tables are stored in the hierarchy `Catalog` -> `Database` -> `Table`. A given table needs at least a database name and a table name specified, as a catalog name can (I think) be inferred.

I thought about it a bit in response to your comment, and I added a step that also raises a `DatasetError` if no database name or no table name is specified. With these changes it is only `validate_catalog` that has a different structure to the other two namespace validation methods. This being the case, my view is that is clearer to keep these two methods separate rather than DRY them.